ArduCopter/Attitude.cpp

```diff
@@ -128,6 +128,12 @@ void Copter::update_throttle_hover()
     if (throttle > 0.0f && abs(climb_rate) < 60 && labs(ahrs.roll_sensor) < 500 && labs(ahrs.pitch_sensor) < 500) {
         // Can we set the time constant automatically
         motors->update_throttle_hover(0.01f);
+        
+        // Learn the wattage required to hover based on current and estimated resting voltage
+        if (g2.hover_watt_learn != HOVER_WATT_LEARN_DISABLED && battery.has_current() && battery.current_amps() > 0 && (float)battery.voltage_resting_estimate() > 0.0 ) {
+            g2.hover_watt = battery.current_amps() * battery.voltage_resting_estimate();
```

Comment:
Do we need some averaging here? Like arithmetical mean of 10-50-100 last samples?

Reply:
YES! I would love to do that. However I am unsure how to accomplish that. Suggestions?

Reply:
A design principle should be that this belongs to the BatteryMon, shouldn't it?

Reply:
@dgrat the reason I didn't put it in the battery or motor library is because hover related functions and parameters only apply to copter. I _could_ split it up between copter and the battery library, but it would actually be twice as much code to do the same thing I think.

Reply:
I think:  `(float)battery.voltage_resting_estimate() > 0.0' is always 'equal battery.voltage_resting_estimate() > 0`

Not sure here, but if abs(battery.current_amps()) > 0: `battery.has_current() && battery.current_amps() > 0` should be always true. on some sensors the measured current can be negative btw.

Reply:
Agree on all of that too. I've actually already cleaned a lot of that up locally and just haven't pushed it yet. I still have a few data type conflicts to clean up though.

Reply:
@Pedals2Paddles for the arithmetical mean of the throttle power, it could be like this
`#define MEAN_SIZE 20
float power_records[MEAN_SIZE];// I'm not sure will that array contain zeroes or a trash at initialization
int power_record_index; // set to 0 in constructor

float MeanPower(float power)
{
if (power_record_index == MEAN_SIZE)
	power_record_index = 0;

power_records[power_record_index++] = power; 

power = 0;
int count = 0;
		
for(int i=0; i<MEAN_SIZE; ++i)
	if (power_records[i] > 0)
	{
		power += power_records[i];
		count++;
	}
		
if (count > 1)
	power /= count;
		
return power;
}`

Reply:
That's generally what I had in mind to write from scratch too.  I was hoping maybe there was some magical pre-existing C++ function that could do it, mainly so I could be lazy.